torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -274,6 +336,22 @@ def to_sharded_dtensor(self, tensor: torch.Tensor) -> DTensor:
             self._global_stride,
         )
 
+    def to_sharded_post_forward_dtensor(self, tensor: torch.Tensor) -> DTensor:
+        if tensor.shape != self.sharded_post_forward_size:
+            _raise_assert_with_print(
+                f"Expects size {self.sharded_post_forward_size} but got {tensor.shape}"
+            )
+        assert self.post_forward_mesh_info is not None  # mypy
+        # TODO: Prefer this DTensor to be read-only and generalize the
```

Comment:
I think the reshard after forward + TP going to be a bit complicated, where it seems like it would fall into a similar case as HSDP + TP?

Reply:
For `reshard_after_forward: int` + TP, since these `DTensor`s will never be used for anything except FSDP all-gather (namely, they will not be used for state dict), would it be possible to only consider the submesh without FSDP's replication?

E.g. if we had 8-way TP and 16-way FSDP with `reshard_after_forward=2`, could we just consider submeshes of shape (2, 8) instead of the full mesh of (8, 2, 8)? If doing that does not provide much benefit, then I am good with handling this case like HSDP + TP.

Reply:
Yes, I think it will be similar to HSDP + TP!